Src/IronPython.Modules/nt.cs

```diff
@@ -355,65 +353,55 @@ public static void chmod(CodeContext context, object? path, int mode, [ParamDict
             }
         }
 
-        private static bool IsValidFd(CodeContext/*!*/ context, int fd) {
-            PythonContext pythonContext = context.LanguageContext;
-            if (pythonContext.FileManager.TryGetFileFromId(pythonContext, fd, out PythonIOModule.FileIO _)) {
-                return true;
-            }
-            if (pythonContext.FileManager.TryGetObjectFromId(pythonContext, fd, out object o)) {
-                if (o is Stream) {
-                    return true;
-                }
-            }
-            return false;
-        }
-
         public static int dup(CodeContext/*!*/ context, int fd) {
-            PythonContext pythonContext = context.LanguageContext;
-            if (pythonContext.FileManager.TryGetFileFromId(pythonContext, fd, out PythonIOModule.FileIO file)) {
-                return pythonContext.FileManager.AddToStrongMapping(file);
+            PythonFileManager fileManager = context.LanguageContext.FileManager;
+
+            object obj = fileManager.GetObjectFromId(fd); // OSError if fd not valid
+            if (obj is PythonIOModule.FileIO file) {
+                var file2 = new PythonIOModule.FileIO(context, file.fileno(context));
+                int fd2 = fileManager.AddFile(file2);
+                fileManager.EnsureRef(file._readStream);
+                fileManager.AddRef(file2._readStream);
+                return fd2;
             } else {
-                Stream? stream = pythonContext.FileManager.GetObjectFromId(fd) as Stream;
-                if (stream == null) {
-                    throw PythonOps.OSError(9, "Bad file descriptor");
-                }
-                return pythonContext.FileManager.AddToStrongMapping(stream);
+                var stream = (Stream)obj;
```

Comment:
It would throw, but currently it is impossible to add to `FileManager` anything else than `FileIO` or `Stream` because only those types are allowed by public API on `PythonFileManager`. If someone tried to add support for any third type, there are a number of places that will need update.

I am not there yet, but with genuine file descriptors, I expect `PythonFileManager` to work quite differently.